mongo/src/org/immutables/mongo/repository/RepositorySetup.java

```diff
@@ -136,7 +182,37 @@ public Builder database(MongoDatabase database) {
      * @return {@code this}
      */
     public Builder gson(Gson gson) {
-      this.gson = checkNotNull(gson);
+      checkNotNull(gson);
+      this.codecRegistry = CodecRegistries.fromRegistries(MongoClient.getDefaultCodecRegistry(),
+              Codecs.codecRegistryFromGson(gson)
+              );
+
+      // legacy naming
+      this.fieldNamingStrategy = new FieldNamingStrategy.GsonNamingStrategy(gson);
+      return this;
+    }
+
+    /**
+     * Register custom codec registry to serialize entities externally.
+     *
+     * @param registry existing registry
+     * @return {@code this}
+     */
+    Builder codecRegistry(CodecRegistry registry) {
+      this.codecRegistry = checkNotNull(registry, "registry");
+      this.fieldNamingStrategy = null;
```

Comment:
troubling as missing up with order of `.fieldNamingStrategy` and `.codecRegistry` could leave it uninialiazed and even if documented, would possibly be one of those wtf moments. btw, this is one more hint to think about somehow merging field naming into codec.

Reply:
Yes. I'm not a big fan of separate `FieldNamingStrategy` myself. Perhaps expose something like:
```java
class Filter  {
    final Member member;
    final Object value;
    final QueryOperators op; // optional  
}
```
This information can be saved during query construction.

Reply:
Than would be definitely moving toward more comprehensive solution, but I think we're still far enough from really optimal solution. Given that we reuse codec registries from mongo driver, I feel this should be acceptable compromise to have separate naming strategy as we currently have in this PR.

I think we just need

* More elaborately explain in Javadoc why naming strategy needed and why it works only for queries, and how codec itself should use the same strategy
* More elaborately explain or at least hint in Javadoc of `translateName` how it supposed to be implemented (the example could be Jackson and that translate name is supposed to read `@JsonProperty` annotation or reach out to Jackson's naming strategies). An example (or hint) is, probably, not required, but nice to have
* Set codec registry and naming strategy at the same time, and in javadoc give yet another reminder and pointer to read strategy's javadoc

```
.codecRegistry(codecRegistry, namingStrategy)
```

Reply:
What would be an optimal solution in your view ? I would like to avoid introducing new interface if there is a better alternative.

So far I see the following options:
1. Explicit `FieldNamingStrategy` interface. I don't like it because it introduces new API.
2. Bson `Encoder<Field/Member/Method>` or `Encoder<Filter>` (where `Filter` is specific class to immutables)

Obviously javadoc has to be improved.

Reply:
I'm kinda ok with `1.` although fully share your dislike it introducing APIs
About `2.` I guess I get where it's heading, but not fully understand it yet, so cannot be sure

Reply:
I've added more explanations to javadoc. Let me know what you think.

Additional question:
In `Repository.generator` [you're generating](https://github.com/immutables/immutables/blob/master/value-processor/src/org/immutables/value/processor/Repositories.generator#L828) separate class with same fields. Any reason why not get members from existing interface or generated immutables class  via same reflection ?

Reply:
as far as I can remember, we cannot always rely on a field to be present, some attributes can be implemented in a special way. In my view, this is just a peculiarity of Gson API to require `Field` object as input. So the current way seems easy and reliable way to get those fields. I've tried to construct some field objects in support library but that was fragile (some com sun unsafe stuff etc). And I don't know if there's a way out of this